### DISCUSSION TITLE
CORS error when using client side NextJS API call for speech-to-text

### DISCUSSION THREAD

**Discussion by michaelmccrae** [deepgram/community#686] [General help]
### Which Deepgram product feature are you using?

Deepgram API - STT Pre-recorded

### Details

I was getting the following error when I was trying to run speech-to-text from NextJS client side app: 

_DeepgramError: Due to CORS we are unable to support REST-based API calls to our API from the browser. Please consider using a proxy, and including a `restProxy: { url: ''}` in your Deepgram client options._

I assume you can only run server side for API calls? I didn't see any information on this error or fixes in Deepgram's docs. 

I am using version 14.01 of NextJS. 

Code I used is below. The **async function query(queryFile)** line is where I take a buffer and try to hit Deepgram API. 

```

'use client';

import * as React from 'react';
import { useState, useEffect } from 'react';
import { AudioRecorder, useAudioRecorder } from 'react-audio-voice-recorder';

// https://developers.deepgram.com/docs/node-sdk-pre-recorded-transcription#transcribing-a-file-buffer

import { createClient } from "@deepgram/sdk";
// const deepgram = createClient("DEEPGRAM_API_KEY");

// import Child from './child';

export default function Hfone() {
    const [blobby, setBlobby] = useState(null);
    

  const recorderControls = useAudioRecorder(
    {
      noiseSuppression: true,
      echoCancellation: true,
    },
    (err) => console.table(err) // onNotAllowedOrFound
  );
  const addAudioElement = (blob) => {
    const url = URL.createObjectURL(blob);
    const audio = document.createElement('audio');
    audio.src = url;
    audio.controls = true;
    document.body.appendChild(audio);
  };

  const {
    startRecording,
    stopRecording,
    togglePauseResume,
    recordingBlob,
    isRecording,
    isPaused,
    recordingTime,
    } = useAudioRecorder();

    

    //  const handleSave = () => {
    //     recorderControls.stopRecording();
    
    //     };


    useEffect(() => {
        if (!recorderControls.recordingBlob) 
        {
            return
        } else {

        let reader = new FileReader(); // no arguments

        console.log("recordingBlob", recorderControls.recordingBlob)

       
        }
    }, [recorderControls.recordingBlob])

    // hf

    // let reader = new FileReader(); // no arguments
    // console.log("recorderControls.recordingBlob in useEffect hf", recorderControls.recordingBlob)
    // console.log("FileReader.readAsArrayBuffer() no await", reader.readAsArrayBuffer(filing))

    async function query(queryFile) {
          console.log("queryFile",queryFile);
          const deepgram = createClient(process.env.NEXT_PUBLIC_DEEPGRAM_API_KEY);
          let reader = new FileReader();

          const { result, error } = await deepgram.listen.prerecorded.transcribeFile(
            // path to the audio file
            reader.readAsArrayBuffer(queryFile),
            // STEP 3: Configure Deepgram options for audio analysis
            {
              model: "nova-2",
              smart_format: true,
            }
          );
        
          if (error) throw error;
          // STEP 4: Print the results
          if (!error) console.dir(result, { depth: null });
        };
        
        
         
        async function transcribeFile() {
            console.log("transcribeFile")
        }

    useEffect(() => {
        // (!filing) ? console.log("no file") : console.log("file", filing)
        if (!recorderControls.recordingBlob) 
        { 
          return
        } else {

            query(recorderControls.recordingBlob)
         
        }

        
    }, [recorderControls.recordingBlob])


    //end hf

  return (
    <div>
      <AudioRecorder
        onRecordingComplete={(blob) => addAudioElement(blob)}
        recorderControls={recorderControls}
    
        // downloadOnSavePress={true}
        // downloadFileExtension="mp3"
        showVisualizer={true}
      />
      <br />
      <button onClick={recorderControls.stopRecording}>Stop recording</button>
      <hr />
      {/* <button onClick={handleSave}>Save</button> */}
     
      <br />
    </div>
  );
}








```

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Comment by michaelmccrae** [comment]
Found part of the answer documented here: https://developers.deepgram.com/docs/node-sdk-pre-recorded-transcription

I don't see any further information/examples on "proxy configuration" in the Deepgram docs. 

_The Deepgram JavaScript SDK now works in both server and browser environments. A proxy configuration is required for browser environments (see the section below)._

**Reply by lukeocodes** [reply]
Information on configuring the SDK for a proxy can be found on the SDK's README.

You'll need to write/provide your own proxy. This can be done as part of the Next.js app, or you can use something stand-alone.

Our REST API does not support client side requests, and therefore we cannot configure CORS to accept requests from your application. This holds a few purposes, but mostly we want clients to responsibly protect their API keys.

Here is an example proxy application I wrote. You could implement something similar.

https://github.com/deepgram-devs/deepgram-node-proxy

You'll have more luck implementing this in a serverless endpoint in a page route, than an app route. Luckily, even the latest Next.js supports both side by side, as long as routes don't clash.

Another alternative is that you use the SDK in the serverless endpoint, and make a regular fetch request to it from the client-side.
